trace/trace.go

```diff
@@ -0,0 +1,111 @@
+package trace
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"regexp"
+
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	maxVersion       = 254
+	supportedVersion = 0
+
+	traceparentHeader = "traceparent"
+	tracestateHeader  = "tracestate"
+)
+
+var traceCtxRegExp = regexp.MustCompile("^(?P<version>[0-9a-f]{2})-(?P<traceID>[a-f0-9]{32})-(?P<spanID>[a-f0-9]{16})-(?P<traceFlags>[a-f0-9]{2})(?:-.*)?$")
+
+// Traceparent get traceparent from spancontext.
+func Traceparent(sc trace.SpanContext) string {
+	flags := sc.TraceFlags() & trace.FlagsSampled
+
+	return fmt.Sprintf("%.2x-%s-%s-%s",
+		supportedVersion,
+		sc.TraceID(),
+		sc.SpanID(),
+		flags)
+}
+
+// ID get traceid from context.
+func ID(ctx context.Context) string {
+	sc := trace.SpanContextFromContext(ctx)
+	if sc.HasTraceID() {
+		return sc.TraceID().String()
+	}
+	return ""
+}
+
+// NewSpanContextFromTrace creates span context.
+func NewSpanContextFromTrace(traceparent, tracestate string) trace.SpanContext {
+	matches := traceCtxRegExp.FindStringSubmatch(traceparent)
+
+	if len(matches) == 0 {
+		return trace.SpanContext{}
+	}
+
+	if len(matches) < 5 { // four subgroups plus the overall match
+		return trace.SpanContext{}
+	}
+
+	if len(matches[1]) != 2 {
+		return trace.SpanContext{}
+	}
+	ver, err := hex.DecodeString(matches[1])
+	if err != nil {
+		return trace.SpanContext{}
+	}
+	version := int(ver[0])
+	if version > maxVersion {
+		return trace.SpanContext{}
+	}
+
+	if version == 0 && len(matches) != 5 { // four subgroups plus the overall match
+		return trace.SpanContext{}
+	}
+
+	if len(matches[2]) != 32 {
```

Comment:
I see lots of magic number, if we cannot make them const, better to add a complete example in the comment.

Reply:
?